python/ray/data/_internal/split.py

```diff
@@ -0,0 +1,242 @@
+        accessor = BlockAccessor.for_block(split_block)
+        split_meta = BlockMetadata(
+            num_rows=accessor.num_rows(),
+            size_bytes=accessor.size_bytes(),
+            schema=meta.schema,
+            input_files=meta.input_files,
+            exec_stats=stats.build(),
+        )
+        split_result.append((ray.put(split_block), split_meta))
+        prev_index = index
+    return (block_id, split_result)
+
+
+def _split_all_blocks(
+    blocks_with_metadata: List[Tuple[ObjectRef[Block], BlockMetadata]],
+    block_rows: List[int],
+    per_block_split_indices: List[List[int]],
+) -> List[Tuple[ObjectRef[Block], BlockMetadata]]:
+    """Split all the input blocks based on the split indices"""
+    split_single_block = cached_remote_fn(_split_single_block)
+
+    all_blocks_split_results: List[List[Tuple[ObjectRef[Block], BlockMetadata]]] = [
+        None
+    ] * len(blocks_with_metadata)
+
+    split_single_block_futures = []
+
+    for block_id, block_split_indices in enumerate(per_block_split_indices):
+        (block_ref, meta) = blocks_with_metadata[block_id]
+        block_row = block_rows[block_id]
+        if len(block_split_indices) == 0:
+            # optimization: if no split is needed, we just need to add it to the
+            # result
+            all_blocks_split_results[block_id] = [(block_ref, meta)]
+        else:
+            # otherwise call split remote function.
+            split_single_block_futures.append(
+                split_single_block.remote(
+                    block_id,
+                    block_ref,
+                    meta,
+                    block_row,
+                    block_split_indices,
+                )
+            )
+    if split_single_block_futures:
+        split_single_block_results = ray.get(split_single_block_futures)
+        for block_id, block_split_result in split_single_block_results:
+            all_blocks_split_results[block_id] = block_split_result
+    return all_blocks_split_results
+
+
+def _generate_global_split_results(
+    all_blocks_split_results: List[List[Tuple[ObjectRef[Block], BlockMetadata]]],
+) -> Tuple[List[List[ObjectRef[Block]]], List[List[BlockMetadata]]]:
+    """Reassemble per block's split result into final split result."""
+    result_blocks = []
+    result_metas = []
+    current_blocks = []
+    current_meta = []
+    for single_block_split_result in all_blocks_split_results:
+        for i, (block, meta) in enumerate(single_block_split_result):
+            # we should create a new global split whenever
+            # we encountered a new local split in the per block
+            # split result.
+            if i != 0:
+                result_blocks.append(current_blocks)
+                result_metas.append(current_meta)
+                current_blocks = []
+                current_meta = []
+            current_blocks.append(block)
+            current_meta.append(meta)
+    result_blocks.append(current_blocks)
```

Comment:
shouldn't we check `current_blocks` is not empty before append?

Reply:
hmm I think empty current_blocks is a valid global split.

Reply:
hmm I may be wrong, but what if last `single_block_split_result` has more than 1 local split, and already going through L171-L174 before this? Why we need to add an extra empty block here?

Reply:
hmm I think you are right. Actually the only case `current_blocks` is empty is because `all_blocks_split_results` is empty, given the invariant each `single_block_split_result` is not empty.